### ISSUE TITLE
CUDA out of memory error

### ISSUE THREAD

**Issue by boazgmail** [allenai/allennlp#2515]
**Describe the bug**
I get Cuda out of memory Error after running my model for a while.
My model is a CVAE.
I am trying to train it on 1,000,000 line file.
I was able to train the model on a smaller file (100,000 lines)  
I am able to read the file to memory.
The model is being trained for a while but does not complete an epoch.
It then crashes with Cuda Out Of Memory error
I tried lazy load and use "biggest_batch_first" : true,

**To Reproduce**
Need to run the model for a while...

**Expected behavior**
I do not expect the model to crash during training as it is running for a while (more than an hour). 

**System (please complete the following information):**
 - OS: Linux
 - Python version: python  3.6.8 h0371630_0
 - AllenNLP version: allennlp 0.8.2-unreleased
 - PyTorch version: (if you installed it yourself)

**Additional context**
Model is running on GPU 0
Error trace:
2019-02-14 10:49:25,824 - INFO - allennlp.training.trainer - GPU 0 memory usage MB: 1340
2019-02-14 10:49:25,825 - INFO - allennlp.training.trainer - GPU 1 memory usage MB: 10608
2019-02-14 10:49:25,825 - INFO - allennlp.training.trainer - Training
  0%|          | 0/1 [00:00<?, ?it/s]2019-02-14 10:49:25,826 - INFO - ctg.dataset_readers.wa_is - Reading instances from lines in file at: data/wa/1000000.tsv
reconstruction: 56.4100, kld: 14.4995, loss: 56.8394 ||: : 3181it [1:22:15, 10.61s/it]Traceback (most recent call last):
  File "/dccstor/knewedge/boazc/anaconda3/envs/py36/lib/python3.6/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/dccstor/knewedge/boazc/anaconda3/envs/py36/lib/python3.6/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/dccstor/knewedge/boazc/anaconda3/envs/py36/lib/python3.6/site-packages/allennlp/run.py", line 18, in <module>
    main(prog="allennlp")
  File "/dccstor/knewedge/boazc/anaconda3/envs/py36/lib/python3.6/site-packages/allennlp/commands/__init__.py", line 101, in main
    args.func(args)
  File "/dccstor/knewedge/boazc/anaconda3/envs/py36/lib/python3.6/site-packages/allennlp/commands/train.py", line 111, in train_model_from_args
    args.force)
  File "/dccstor/knewedge/boazc/anaconda3/envs/py36/lib/python3.6/site-packages/allennlp/commands/train.py", line 144, in train_model_from_file
    return train_model(params, serialization_dir, file_friendly_logging, recover, force)
  File "/dccstor/knewedge/boazc/anaconda3/envs/py36/lib/python3.6/site-packages/allennlp/commands/train.py", line 366, in train_model
    metrics = trainer.train()
  File "/dccstor/knewedge/boazc/anaconda3/envs/py36/lib/python3.6/site-packages/allennlp/training/trainer.py", line 823, in train
    train_metrics = self._train_epoch(epoch)
  File "/dccstor/knewedge/boazc/anaconda3/envs/py36/lib/python3.6/site-packages/allennlp/training/trainer.py", line 530, in _train_epoch
    loss.backward()
  File "/dccstor/knewedge/boazc/anaconda3/envs/py36/lib/python3.6/site-packages/torch/tensor.py", line 102, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph)
  File "/dccstor/knewedge/boazc/anaconda3/envs/py36/lib/python3.6/site-packages/torch/autograd/__init__.py", line 90, in backward
    allow_unreachable=True)  # allow_unreachable flag
RuntimeError: CUDA out of memory. Tried to allocate 2.25 GiB (GPU 0; 11.17 GiB total capacity; 6.48 GiB already allocated; 2.23 GiB free; 2.12 GiB cached)

Model is running on GPU 0
nvidia-smi:
Thu Feb 14 12:07:52 2019       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 387.26                 Driver Version: 387.26                    |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|===============================+======================+=================|
|   0  Tesla K40m          On   | 00000000:1B:00.0 Off |                    0 |
| N/A   48C    P0    78W / 235W |  11084MiB / 11439MiB |     76%   E. Process |
+-------------------------------+----------------------+----------------------+
|   1  Tesla K40m          On   | 00000000:86:00.0 Off |                    0 |
| N/A   57C    P0   143W / 235W |  10608MiB / 11439MiB |     95%   E. Process |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                       GPU Memory |
|  GPU       PID   Type   Process name                             Usage      |
|=======================================================================|
|    0     16928      C   python                                     11070MiB |
|    1     23883      C   python                                     10594MiB |
+-----------------------------------------------------------------------------+

jsonnet:
{
  "random_seed": SEED,
  "numpy_seed": SEED,
  "pytorch_seed": SEED,
  "dataset_reader": {
    "type": READER,
    "use_conditions": USE_CONDITIONS
  },
  "train_data_path": DATA_PATH + "/train_wa.txt",
  "validation_data_path": DATA_PATH + "/valid_wa.txt",
  "model": {
    "type": "cvae_is",
    "source_encoder": {
      "type": "masked_encoder",
      "source_embedder": {
        "token_embedders": {
          "tokens": {
            "type": "embedding",
            "vocab_namespace": "tokens",
            "embedding_dim": EMBEDDING_DIM
          }
        }
      },
      "rnn": {
        "type": "lstm",
        "input_size": EMBEDDING_DIM,
        "hidden_size": EMBEDDING_DIM,
      }
    },
    "condition_encoder": {
      "type": "masked_encoder",
      "source_embedder": {
        "token_embedders": {
          "tokens": {
            "type": "embedding",
            "vocab_namespace": "tokens",
            "embedding_dim": EMBEDDING_DIM
          }
        }
      },
      "rnn": {
        "type": "lstm",
        "input_size": EMBEDDING_DIM,
        "hidden_size": CONDITION_DIM,
      }
    },
    "decoder": {
      "type": "variational_decoder",
      "target_embedder": {
        "token_embedders": {
          "tokens": {
            "type": "embedding",
            "vocab_namespace": "tokens",
            "embedding_dim": EMBEDDING_DIM
          }
        }
      },
      "rnn": {
        "type": "lstm",
        "input_size": EMBEDDING_DIM + LATENT_DIM + CONDITION_DIM,
        "hidden_size": HIDDEN_DIM,
      },
      "latent_dim": LATENT_DIM + CONDITION_DIM,
      "max_decoding_steps": MAX_DECODING_STEPS,
      "beam_size": BEAM_SIZE,
    },
    "latent_dim": LATENT_DIM,
    "initializer": [
      [".*", {"type": "uniform", "a": -INIT_UNIFORM_RANGE_AROUND_ZERO, "b": INIT_UNIFORM_RANGE_AROUND_ZERO}]
    ]
  },
  "iterator": {
    "type": "bucket",
    "batch_size" : BATCH_SIZE,
    "biggest_batch_first" : true,
    "sorting_keys": [["source", "num_tokens"]]
  },
  "trainer": {
    "num_epochs": NUM_EPOCHS,
    "patience": NUM_EPOCHS,
    "cuda_device": CUDA,
    "summary_interval": SUMMARY_INTERVAL,
    "grad_clipping": GRAD_CLIPPING,
    "grad_norm": GRAD_NORM,
    "loss_weights": {
      "reconstruction": {
        "initial_weight": RECONSTRUCTION_WEIGHT
      },
      "kld": {
        "type": "linear_annealer",
        "min_weight": ANNEAL_MIN_WEIGHT,
        "max_weight": ANNEAL_MAX_WEIGHT,
        "warmup": ANNEAL_WARMUP,
        "num_iter_to_max": ANNEAL_NUM_ITER_TO_MAX,
      },
    },
    "should_log_parameter_statistics": SHOULD_LOG_PARAMETER_STATISTICS,
    "should_log_learning_rate": SHOULD_LOG_LEARNING_RATE,
    "optimizer": {
      "type": OPTIMIZER,
      "lr": LEARNING_RATE
    }
  }
}

**Comment by schmmd** [comment]
@boazgmail is CVAE a model you implemented?  If so, I'm not sure how we'd troubleshoot this effectively with the information you've given us.

**Comment by boazgmail** [comment]
Thanks for looking into it. I understand the difficulty to troubleshoot. The code is based on a vae pull request that my colleague submitted coupe of weeks ago, thus, you may be able to take a look and give it a try. But please note that I do not think the problem is in the code. I suspect it relates to the large vocabulary size (~200K tokens) and unfortunately I will not be able to share the data. I am now try to limit the vocabulary size and trace results...

**Comment by mynameisguy** [comment]
had the same issue, with Allennlp model.
working on WMT14 dataset eng->de (1,536,168 lines in train).
creating vocab using create-vocab with the size of 32,000 of word pieces. in the middle of the first epoch, it crashes.

config:
```
{
    "dataset_reader": {
        "type": "seq2seq",
        "lazy": false,
        "source_tokenizer": {
            "type": "word",
            "word_splitter": {
                "type": "bert-basic"
            }
        },
        "target_tokenizer": {
            "type": "word",
            "word_splitter": {
                "type": "bert-basic"
            }
        }
    },
    "iterator": {
        "type": "bucket",
        "batch_size": 32,
        "sorting_keys": [
            [
                "source_tokens",
                "num_tokens"
            ],
            [
                "target_tokens",
                "num_tokens"
            ]
        ]
    },
    "model": {
        "type": "simple_seq2seq",
        "attention": {
            "type": "dot_product"
        },
        "beam_size": 8,
        "encoder": {
            "type": "stacked_self_attention",
            "dropout_prob": 0.1,
            "feedforward_hidden_dim": 2048,
            "hidden_dim": 512,
            "input_dim": 512,
            "num_attention_heads": 8,
            "num_layers": 6,
            "projection_dim": 32,
            "use_positional_encoding": true
        },
        "max_decoding_steps": 150,
        "source_embedder": {
            "type": "basic",
            "token_embedders": {
                "tokens": {
                    "type": "embedding",
                    "embedding_dim": 512,
                    "trainable": true
                }
            }
        }
    },
    "train_data_path": "data/WMT14/europarl_eng_to_de_train.txt",
    "validation_data_path": "data/WMT14/europarl_eng_to_de_valid.txt",
    "trainer": {
        "cuda_device": 2,
        "grad_clipping": 5,
        "num_epochs": 40,
        "optimizer": {
            "type": "adam",
            "lr": 0.001
        },
        "patience": 10,
        "should_log_learning_rate": true,
        "should_log_parameter_statistics": true,
        "validation_metric": "+BLEU"
    },
    "vocabulary": {
        "directory_path": "results/mt/vocab/vocabulary"
    }
}
```

the error:
```
le "/home/user/miniconda3/lib/python3.7/site-packages/allennlp/training/trainer.py", line 465, in train
    train_metrics = self._train_epoch(epoch)
  File "/home/user/miniconda3/lib/python3.7/site-packages/allennlp/training/trainer.py", line 306, in _train_epoch
    loss = self.batch_loss(batch_group, for_training=True)
  File "/home/user/miniconda3/lib/python3.7/site-packages/allennlp/training/trainer.py", line 247, in batch_loss
    output_dict = self.model(**batch)
  File "/home/user/miniconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 489, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/user/miniconda3/lib/python3.7/site-packages/allennlp/models/encoder_decoders/simple_seq2seq.py", line 212, in forward
    state = self._encode(source_tokens)
  File "/home/user/miniconda3/lib/python3.7/site-packages/allennlp/models/encoder_decoders/simple_seq2seq.py", line 272, in _encode
    encoder_outputs = self._encoder(embedded_input, source_mask)
  File "/home/user/miniconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 489, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/user/miniconda3/lib/python3.7/site-packages/allennlp/modules/seq2seq_encoders/stacked_self_attention.py", line 150, in forward
    attention_output = attention(feedforward_output, mask)
  File "/home/user/miniconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 489, in __call__
    result = self.forward(*input, **kwargs)
  File "/home/user/miniconda3/lib/python3.7/site-packages/allennlp/modules/seq2seq_encoders/multi_head_self_attention.py", line 133, in forward
    memory_efficient=True)
  File "/home/user/miniconda3/lib/python3.7/site-packages/allennlp/nn/util.py", line 267, in masked_softmax
    masked_vector = vector.masked_fill((1 - mask).byte(), mask_fill_value)
  File "/home/user/miniconda3/lib/python3.7/site-packages/torch/tensor.py", line 337, in masked_fill
    return self.clone().masked_fill_(mask, value)
RuntimeError: CUDA out of memory. Tried to allocate 288.00 MiB (GPU 2; 10.73 GiB total capacity; 9.43 GiB already allocated; 197.62 MiB free; 258.96 MiB cached)
```

running allennlp 0.8.2 on ubunutu, installed using pip

**Comment by boazgmail** [comment]
I did many runs and concluded that the size of the vocabulary is the major contribution to this OOM error. @schmmd - maybe you want to take another look on a different model with large vocabulary to try and figure this out

**Comment by schmmd** [comment]
Closing this issue since @boazgmail came to a conclusion about the root cause.  Unfortunately I don't think we'll have time to root cause it further, although these details might be helpful for #2569.

**Comment by YongfeiYan** [comment]
I encountered a CUDA Out of Memory Error like you, with a vocabulary 10k. 
I found that I forgot to limit the sequence in the reader, which was problematic if a vary long sentence appeared, like 500. CUDA memory stays constant when I set `max_sequence_length=50`.
BTW, if CUDA memory keeps going during training, the computation graph may be not released, e.g. accumulating loss in metrics instead of using loss.item()
